### PR TITLE
Compat: Refactor vertex_state tests for new limits

### PR DESCRIPTION
**Pull request by greggman**
Compat has different limits than non-Compat and tons of the CTS needs to be refactored from assuming the default limits are fixed.

Note: Locally I have `kLimitInfo` deleted as well as  `kMaxColorAttachments`, `kMaxVertexBuffers `, `kMaxVertexAttributes`, `kMaxVertexBufferArrayStride`

But, deleting those meant changing 35 files with 2500 lines of changes so this is just the first file to keep the PRs small.

I'll try upload one file at a time and then finally delete those constants.

Note 2: I couldn't come up with a better name for `SpecValue`. I'm happy to change it but it would be nice if whatever it becomes was short because the longer it gets the harder the code becomes to read as things get reformatted across multiple lines.

Note 3: Right now, kLimitInfoCore and kLimitInfoCompat are generated with 2 calls to `makeTableWithDefaults`. Theoretically both could be created with one call to something but it was taking too long for me to figure out the magic type incantation so to move forward I just did the simplested thing I could think of. It doesn't leak outside of capabilities_info.ts so I'm hoping that can be fixed later.

<hr>

**Requirements for PR author:**

- [X] All missing test coverage is tracked with "TODO" or `.unimplemented()`.
- [X] New helpers are `/** documented */` and new helper files are found in `helper_index.txt`.
- [X] Test behaves as expected in a WebGPU implementation. (If not passing, explain above.)

**Requirements for [reviewer sign-off](https://github.com/gpuweb/cts/blob/main/docs/reviews.md):**

- [ ] Tests are properly located in the test tree.
- [ ] [Test descriptions](https://github.com/gpuweb/cts/blob/main/docs/intro/plans.md) allow a reader to "read only the test plans and evaluate coverage completeness", and accurately reflect the test code.
- [ ] Tests provide complete coverage (including validation control cases). **Missing coverage MUST be covered by TODOs.**
- [ ] Helpers and types promote readability and maintainability.

When landing this PR, be sure to make any necessary issue status updates.

### CODE REVIEW

File: src/common/util/util.ts [hunk 1]
```diff
@@ -423,3 +423,29 @@ export function memcpy(
 ): void {
   subarrayAsU8(dst.dst, dst).set(subarrayAsU8(src.src, src));
 }
+
+/**
+ * Used to create a value that is specified by multiplying some runtime value
+ * by a constant and then adding a constant to it.
+ */
+export interface SpecValue {
```
Comment (Kangz) [inline src/common/util/util.ts:431]: This name seems too general for what it is, but I don't really have a better idea. `ValueTestVariant` maybe?

**Review by Kangz — APPROVED** [review]

File: src/common/util/data_tables.ts [hunk 2]
```diff
@@ -51,3 +51,66 @@ export function makeTable<
...
+
+/**
+ * Creates an info lookup object from a more nicely-formatted table.
+ *
+ * Note: Using `as const` on the arguments to this function is necessary to infer the correct type.
+ *
+ * Example:
+ *
+ * ```
+ * const t = makeTableWithDefaults(
```
[the remark this answers is not in the document]
Comment (greggman) [reply src/common/util/data_tables.ts:63]: good idea. Done. Another idea is just to pass in a full remapping as in `{a: 'a', 'c': default, d: 'd' }` but I don't know how to make the type stuff happy with that. I put in a maintenance_todo about it. Fortunately it's only used one place ATM.

File: src/webgpu/gpu_test.ts [hunk 3]
```diff
@@ -200,6 +206,13 @@ export class GPUTestSubcaseBatchState extends SubcaseBatchState {
     throw new SkipTestCase(msg);
   }
 
+  /** Throws an exception making the subcase as skipped if condition is true */
+  skipIf(cond: boolean, msg: string = '') {
```
[the remark this answers is not in the document]
Comment (greggman) [reply src/webgpu/gpu_test.ts:210]: done

File: src/webgpu/capability_info.ts [hunk 4]
```diff
@@ -642,50 +649,101 @@ export const kIndexFormat: readonly GPUIndexFormat[] = ['uint16', 'uint32'];
...
+  'maxComputeWorkgroupSizeX':                  [           ,       256,             128,                          ],
+  'maxComputeWorkgroupSizeY':                  [           ,       256,             128,                          ],
+  'maxComputeWorkgroupSizeZ':                  [           ,        64,              64,                          ],
+  'maxComputeWorkgroupsPerDimension':          [           ,     65535,           65535,                          ],
+} as const];
+
+export const kFeatureLevels = ['core', 'compatibility'] as const;
+export type FeatureLevel = typeof kFeatureLevels[number];
+
+const kLimitKeys = ['class', 'default', 'maximumValue'] as const;
```
[the remark this answers is not in the document]
Comment (greggman) [reply src/webgpu/capability_info.ts:700]: done

File: src/common/util/util.ts (hunk 1 above)
Comment (greggman) [reply src/common/util/util.ts:431]: done

File: src/common/util/data_tables.ts [hunk 5]
```diff
@@ -51,3 +51,66 @@ export function makeTable<
...
+  Defaults extends readonly unknown[],
+  Table extends { readonly [k: string]: readonly unknown[] }
+>(
+  defaultName: string,
+  members: Members,
+  dataMembers: DataMembers,
+  defaults: Defaults,
+  table: Table
+): {
+  readonly [k in keyof Table]: ResolveType<ZipKeysWithValues<Members, Table[k], Defaults>>;
```
[the remark this answers is not in the document]
Comment (greggman) [reply src/common/util/data_tables.ts:110]: will address

**Pull request merged**